example/tests/pin_api_test.py

```diff
@@ -0,0 +1,209 @@
+  p = origen.dut.add_pin("test_pin")
+  assert isinstance(p, _origen.dut.pins.Pin)
+  assert len(origen.dut.pins) == 1
+  p2 = origen.dut.pins['test_pin']
+  assert isinstance(p2, _origen.dut.pins.Pin)
+  p3 = origen.dut.pin('test_pin')
+  assert isinstance(p3, _origen.dut.pins.Pin)
+  #assert 'test_pin' in origen.dut.pins # origen.dut.has_pin('test_pin')
+
+  # Check the state of the pin itself
+  # This should be the default state
+  assert p.name == 'test_pin'
+  assert p._path == ''
+  assert p.data == 0
+  assert p.action == "HighZ"
+
+  # Add another pin
+  p = origen.dut.add_pin("other_pin")
+  assert isinstance(p, _origen.dut.pins.Pin)
+  assert len(origen.dut.pins) == 2
+  assert p.name == origen.dut.pins['other_pin'].name
+  assert p.name == origen.dut.pin('other_pin').name
+  #assert origen.dut.has_pin('other_pin')
+
+  # Attempt to posture the pin.
+  assert p.data == 0
+  p.data = 1
+  assert p.data == 1
+  p.set(0)
+  assert p.data == 0
+
+  # Drive/Verify/Capture Pins
+  assert p.action == "HighZ"
+  p.drive()
+  assert p.data == 0
+  assert p.action == "Drive"
+  p.verify()
+  assert p.data == 0
+  assert p.action == "Verify"
+  p.capture()
+  assert p.data == 0
+  assert p.action == "Capture"
+  p.highz()
+  assert p.data == 0
+  assert p.action == "HighZ"
+
+  p.drive(1)
+  assert p.data == 1
+  assert p.action == "Drive"
+  p.verify(0)
+  assert p.data == 0
+  assert p.action == "Verify"
+  p.capture()
+  assert p.data == 0
+  assert p.action == "Capture"
+  p.highz()
+  assert p.action == "HighZ"
+
+  # Some cases to ensure errors and bad input are handled.
+  # Access an unknown pin
+  with pytest.raises(KeyError):
+    origen.dut.pins['blah']
+  with pytest.raises(OSError):
+    # Ensure we cannot add the same pin over again.
+    origen.dut.add_pin("test_pin")
+  with pytest.raises(OSError):
+    p.data = 2
+  with pytest.raises(OSError):
+    p.drive(2)
+  with pytest.raises(OSError):
+    p.verify(2)
+  with pytest.raises(OSError):
+    p.set(2)
+  assert p.action == "HighZ"
+
+  # Check that pins are available on subblocks
+  assert len(origen.dut.pins) == 2
+  assert origen.dut.sub_blocks["core1"]
+  assert len(origen.dut.sub_blocks["core1"].pins) == 0
+  assert origen.dut.sub_blocks["core1"].add_pin("test_pin_core1")
+  assert len(origen.dut.sub_blocks["core1"].pins) == 1
+  p = origen.dut.sub_blocks["core1"].pin("test_pin_core1")
+  assert p._path == "core1"
+  assert len(origen.dut.pins) == 2
+
+  # Add a pin alias
+  p = origen.dut.pin('test_pin')
+  #p.add_alias('test_alias')
+  origen.dut.add_pin_alias('test_pin', 'test_alias1')
+  alias_p = origen.dut.pin('test_alias1')
+  assert alias_p.name == p.name
+  assert len(origen.dut.pins) == 2
+
+  # Add multiple aliases at once.
+  origen.dut.add_pin_alias('test_pin', 'test_alias2', 'test_alias3')
+  assert len(origen.dut.pins) == 2
+  assert origen.dut.pin('test_alias2').name == 'test_pin'
+  assert origen.dut.pin('test_alias3').name == 'test_pin'
```

Comment:
Should we use .id for the primary identifier rather than name?
On O1 we weren't really consistent on this and used both, for O2 I have been trying to use id instead of name.

Reply:
I'm fine with `id`. I always used `name` in O1 so I went with that. No other reason. I can make that change.

Reply:
Updated getters and parameter names to 'id'.